### PR TITLE
plan: flatten row op row (op = GE, GT, LE, LT) in expression_rewriter

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit flatten `row op row` (op = GE, GT, LE, LT) in expression_rewriter

Before this commit, we only convert `row op row`(op = NE, EQ, NullEQ) to a CNF.
e.g. `(a0,a1,a2) op (b0,b1,b2)` will be converted to `(a0 op b0) and (a1 op b1) and (a2 op b2)`

After this commit, op `GE`, `GT`, `LE`, `LT` can also be flattened.
If op are LE or GE, `(a0,a1,a2) op (b0,b1,b2)` will be converted to
``` sql
IF( (a0 op b0) EQ 0, 0, 
    IF((a1 op b1) EQ 0, 0, a2 op b2))
```
If op are LT or GT, `(a0,a1,a2) op (b0,b1,b2)` will be converted to
``` sql
IF(a0 NE b0, a0 op b0,
     IF(a1 NE b1,
         a1 op b1,
         a2 op b2
     )
 )
```

PTAL @hanfei1991 @winoros @shenli

### CODE REVIEW

File: plan/resolve_indices.go [hunk 1]
```diff
@@ -14,13 +14,15 @@
...
 import (
+	"github.com/ngaut/log"
 	"github.com/pingcap/tidb/expression"
 )
 
 // ResolveIndices implements Plan interface.
 func (p *Projection) ResolveIndices() {
 	p.basePlan.ResolveIndices()
 	for _, expr := range p.Exprs {
+		log.Warning(expr.String())
```
Comment (hanfei1991) [inline plan/resolve_indices.go]: remove debug log

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -131,23 +131,71 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
...
+	}
+	c, _ := e.(*expression.Constant)
+	ret = &expression.Constant{Value: types.NewDatum(c.Value.GetRow()[1:]), RetType: c.GetType()}
+	return
+}
+
+// 1. If op are EQ or NE or NullEQ, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0) and (a1 op b1) and (a2 op b2)
+// 2. If op are LE or GE, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0)
+// 3. If op are LT or GT, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to
+// `CASE a0 EQ b0
```
Comment (hanfei1991) [inline plan/expression_rewriter.go]: I think `if` is much clearer...

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @zimulala

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -131,23 +131,69 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
...
+	}
+	c, _ := e.(*expression.Constant)
+	ret = &expression.Constant{Value: types.NewDatum(c.Value.GetRow()[1:]), RetType: c.GetType()}
+	return
+}
+
+// 1. If op are EQ or NE or NullEQ, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0) and (a1 op b1) and (a2 op b2)
+// 2. If op are LE or GE, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0)
+// 3. If op are LT or GT, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to
+// `IF((a0 EQ b0) EQ 0,
```
Comment (hanfei1991) [inline plan/expression_rewriter.go]: (a0 EQ b0) EQ 0 is equal to a0 NE b0 ...

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @shenli

**Comment by shenli** [comment]
Please add a few test cases.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -131,23 +131,72 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
...
+		return ret, errors.Trace(err)
+	}
+	c, _ := e.(*expression.Constant)
+	ret = &expression.Constant{Value: types.NewDatum(c.Value.GetRow()[1:]), RetType: c.GetType()}
+	return
+}
+
+// 1. If op are EQ or NE or NullEQ, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0) and (a1 op b1) and (a2 op b2)
+// 2. If op are LE or GE, constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to
+// `IF( a0 op b0, 1,
```
Comment (shenli) [inline plan/expression_rewriter.go]: The description is not right.

File: plan/expression_rewriter.go [hunk 5]
```diff
@@ -131,23 +131,72 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
 	return &expression.Constant{Value: d, RetType: c.GetType()}
 }
 
-// constructBinaryOpFunctions converts (a0,a1,a2) op (b0,b1,b2) to (a0 op b0) and (a1 op b1) and (a2 op b2).
+func popRowArg(ctx context.Context, e expression.Expression) (ret expression.Expression, err error) {
```
Comment (shenli) [inline plan/expression_rewriter.go]: Add comment for this function.

File: plan/expression_rewriter.go [hunk 6]
```diff
@@ -131,23 +131,72 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
...
+			funcs[i], err = er.constructBinaryOpFunction(getRowArg(l, i), getRowArg(r, i), op)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		return expression.ComposeCNFCondition(er.ctx, funcs...), nil
+	default:
+		larg0, rarg0 := getRowArg(l, 0), getRowArg(r, 0)
+		var expr1, expr2, expr3 expression.Expression
+		if op == ast.LE && op == ast.GE {
```
Comment (shenli) [inline plan/expression_rewriter.go]: Will there be other ops?

Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: I've confirmed it that, there will not be other ops.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

**Comment by shenli** [comment]
@XuHuaiyu Please update the description.

**Comment by XuHuaiyu** [comment]
fixed @shenli

**Pull request merged**